lightning/src/ln/peers/handshake/states.rs

```diff
@@ -1,30 +1,888 @@
+	responder_static_private_key: SecretKey,
+	responder_ephemeral_private_key: SecretKey,
+	responder_ephemeral_public_key: PublicKey,
+	chaining_key: Sha256,
+	hash: Sha256,
+	act_one_builder: ActBuilder
+}
+
+// Handshake state of the Initiator prior to receiving Act 2
+pub(super) struct InitiatorAwaitingActTwoState {
+	initiator_static_private_key: SecretKey,
+	initiator_static_public_key: PublicKey,
+	initiator_ephemeral_private_key: SecretKey,
+	responder_static_public_key: PublicKey,
+	chaining_key: ChainingKey,
+	hash: Sha256,
+	act_two_builder: ActBuilder
+}
+
+// Handshake state of the Responder prior to receiving Act 3
+pub(super) struct ResponderAwaitingActThreeState {
+	hash: Sha256,
+	responder_ephemeral_private_key: SecretKey,
+	chaining_key: ChainingKey,
+	temporary_key: [u8; 32],
+	act_three_builder: ActBuilder
+}
+
+impl InitiatorStartingState {
+	pub(crate) fn new(initiator_static_private_key: SecretKey, initiator_ephemeral_private_key: SecretKey, responder_static_public_key: PublicKey) -> Self {
+		let initiator_static_public_key = private_key_to_public_key(&initiator_static_private_key);
+		let (hash, chaining_key) = handshake_state_initialization(&responder_static_public_key);
+		let initiator_ephemeral_public_key = private_key_to_public_key(&initiator_ephemeral_private_key);
+		InitiatorStartingState {
+			initiator_static_private_key,
+			initiator_static_public_key,
+			initiator_ephemeral_private_key,
+			initiator_ephemeral_public_key,
+			responder_static_public_key,
+			chaining_key,
+			hash
+		}
+	}
+}
+
+impl IHandshakeState for InitiatorStartingState {
+	// https://github.com/lightningnetwork/lightning-rfc/blob/master/08-transport.md#act-one (sender)
+	fn next(self, input: &[u8]) -> Result<(Option<Act>, HandshakeState), String> {
+
+		if input.len() > 0 {
+			return Err("first call for initiator must be empty".to_string());
+		}
+
+		let initiator_static_private_key = self.initiator_static_private_key;
+		let initiator_static_public_key = self.initiator_static_public_key;
+		let initiator_ephemeral_private_key = self.initiator_ephemeral_private_key;
+		let initiator_ephemeral_public_key = self.initiator_ephemeral_public_key;
+		let responder_static_public_key = self.responder_static_public_key;
+		let chaining_key = self.chaining_key;
+		let hash = self.hash;
+
+		// serialize act one
+		let mut act_one = EMPTY_ACT_ONE;
+		let (hash, chaining_key, _) = calculate_act_message(
+			&initiator_ephemeral_private_key,
+			&initiator_ephemeral_public_key,
+			&responder_static_public_key,
+			chaining_key.into_inner(),
+			hash,
+			&mut act_one
+		);
+
+		Ok((
+			Some(Act::One(act_one)),
+			HandshakeState::InitiatorAwaitingActTwo(InitiatorAwaitingActTwoState {
+				initiator_static_private_key,
+				initiator_static_public_key,
+				initiator_ephemeral_private_key,
+				responder_static_public_key,
+				chaining_key,
+				hash,
+				act_two_builder: ActBuilder::new(Act::Two(EMPTY_ACT_TWO))
```

Comment:
I wonder, if the subsequent act builders are always initialized empty, is there a reason to instantiate them here, rather than once the act information comes in? My guess would be this approach obviates the distinction between partial and complete act inputs, as either way, the builder is already initialized

Reply:
Ya, the logic is easier if we don't have to conditionally figure it out later. And, this is just a simple convenience wrapper around a byte array which doesn't really call for a lot of instantiation antics.